geoh5py/workspace/workspace.py

```diff
@@ -955,14 +919,24 @@ def workspace(self):
         """
         return self
 
-    def _io_call(self, file, fun, *args, **kwargs):
+    def _io_call(self, fun, *args, mode="r", **kwargs):
         """
         Run a H5Writer or H5Reader function with validation of target geoh5
         """
-        with fetch_h5_handle(self.validate_file(file)) as h5file:
-            result = fun(h5file, *args, **kwargs)
+        if mode in ["r+", "a"] and self.geoh5.mode == "r":
+            raise UserWarning(
+                f"Error performing {fun}. "
+                "Attempting to write to a geoh5 file in read-only mode. "
+                "Consider closing the workspace (Geoscience ANALYST) and "
+                "re-opening in mode='r+'."
+            )
+        return fun(self.geoh5, *args, **kwargs)
+
+    def __enter__(self):
+        return self
 
-        return result
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.close()
```

Comment:
**suggestion:** make the class inherit from AbstractContextManager,
- it states that is is a context manager, usable in with statement
- it provide a default implementation for `__enter__` (the same as yours 😉 )
See https://docs.python.org/3/library/contextlib.html#contextlib.AbstractContextManager

Alternatively, do not make it a ContextManager (no `__enter__` nor `__exit__`, but use
`with closing(Workspace(file)) as workspace:`.
See https://docs.python.org/3/library/contextlib.html#contextlib.closing

Reply:
Added AbstractContextManager on workspace. Done